File: backend/app/schemas/recette_tco.py
"""Schemas Pydantic pour RecetteTCO"""
from pydantic import BaseModel, ConfigDict
from typing import Optional, List


class RecetteTCOLigneBase(BaseModel):
    id_produit: int
    quantite:   float
    unite:      Optional[str] = None   # mL | L | g | Kg
    note_ligne: Optional[str] = None
    ordre:      int = 0


class RecetteTCOLigneCreate(RecetteTCOLigneBase):
    pass


class RecetteTCOLigneRead(RecetteTCOLigneBase):
    id_ligne:     int
    nom_produit:  Optional[str] = None   # enrichi depuis ProduitEngrais
    type_produit: Optional[str] = None   # enrichi depuis ProduitEngrais
    model_config  = ConfigDict(from_attributes=True)


# ── Recette principale ────────────────────────────────────────────────────────

class RecetteTCOBase(BaseModel):
    nom_recette:         str
    type_tco:            Optional[str]  = None   # Croissance | Stretch | Floraison | Correctif
    quantite_tco:        Optional[float]= None
    unite_tco:           Optional[str]  = None   # L | mL
    duree_oxygenation_h: Optional[int]  = None   # heures d'oxygénation avant utilisation
    notes:               Optional[str]  = None


class RecetteTCOCreate(RecetteTCOBase):
    lignes: List[RecetteTCOLigneCreate] = []


class RecetteTCOUpdate(BaseModel):
    nom_recette:         Optional[str]  = None
    type_tco:            Optional[str]  = None
    quantite_tco:        Optional[float]= None
    unite_tco:           Optional[str]  = None
    duree_oxygenation_h: Optional[int]  = None
    notes:               Optional[str]  = None
    lignes:              Optional[List[RecetteTCOLigneCreate]] = None


class RecetteTCORead(RecetteTCOBase):
    id_recette_tco: int
    lignes:         List[RecetteTCOLigneRead] = []
    model_config    = ConfigDict(from_attributes=True)

File: frontend/src/api/recetteTCO.ts
import client from './client'

export interface RecetteTCOLigne {
  id_ligne?:    number
  id_produit:   number
  quantite:     number
  unite?:       string   // mL | L | g | Kg
  note_ligne?:  string
  ordre:        number
  // enrichis depuis le backend
  nom_produit?: string
  type_produit?: string
}

export interface RecetteTCO {
  id_recette_tco:       number
  nom_recette:          string
  type_tco?:            string   // Croissance | Stretch | Floraison | Correctif
  quantite_tco?:        number
  unite_tco?:           string   // L | mL
  duree_oxygenation_h?: number   // heures d'oxygénation avant utilisation
  notes?:               string
  lignes:               RecetteTCOLigne[]
}

export type RecetteTCOCreate = Omit<RecetteTCO, 'id_recette_tco'>
export type RecetteTCOUpdate = Partial<RecetteTCOCreate>

export const recetteTCOAPI = {
  getAll:    ()                                           => client.get<RecetteTCO[]>('/recettes-tco/'),
  getById:   (id: number)                                 => client.get<RecetteTCO>(`/recettes-tco/${id}`),
  create:    (data: RecetteTCOCreate)                     => client.post<RecetteTCO>('/recettes-tco/', data),
  update:    (id: number, data: RecetteTCOUpdate)         => client.put<RecetteTCO>(`/recettes-tco/${id}`, data),
  delete:    (id: number)                                 => client.delete(`/recettes-tco/${id}`),
  exportCSV: ()                                           => client.get('/recettes-tco/export/csv', { responseType: 'blob' }),
  importCSV: (file: File) => {
    const form = new FormData(); form.append('file', file)
    return client.post('/recettes-tco/import', form, { headers: { 'Content-Type': 'multipart/form-data' } })
  },
}
